File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys
        

def solution():
  n = int(sys.stdin.readline())
  dp = []
  for i in range(n):
    dp.append([0] * (i + 1))
  inputs = []
  for _ in range(n):
    inputs.append(list(map(int, sys.stdin.readline().split())))
  dp[0][0] = inputs[0][0]
  for i in range(1, n):
    for j in range(i + 1):
      if j == 0:
        dp[i][j] = dp[i - 1][j] + inputs[i][j]
        continue
      if j == i:
        dp[i][j] = dp[i - 1][j - 1] + inputs[i][j]
        continue
      dp[i][j] = max(dp[i - 1][j - 1] + inputs[i][j], dp[i - 1][j] + inputs[i][j])
  print(max(dp[n - 1]))

solution()

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.js
let fs = require('fs');
let inputs = fs.readFileSync('/dev/stdin').toString().split('\n');
const n = Number(inputs[0]);
const dp = [[Number(inputs[1])]];

for (let i = 2; i <= n; i++) {
    const newLine = [];
    const numbers = inputs[i].split(' ').map(Number);
    for (let j = 0; j < numbers.length; j++) {
        if (j === 0) {
            newLine.push(dp[dp.length - 1][j] + numbers[j]);
        } else if (j === numbers.length - 1) {
            newLine.push(dp[dp.length - 1][j - 1] + numbers[j]);
        } else {
            newLine.push(Math.max(dp[dp.length - 1][j - 1], dp[dp.length - 1][j]) + numbers[j]);
        }
    }
    dp.push(newLine);
}
console.log(Math.max(...dp[dp.length - 1]));